File: server/src/main/kotlin/pl/edu/agh/eaiib/io/sp/service/ProcessorService.java
package pl.edu.agh.eaiib.io.sp.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import pl.edu.agh.eaiib.io.sp.common.model.ProcessedData;
import pl.edu.agh.eaiib.io.sp.common.model.Reading;
import pl.edu.agh.eaiib.io.sp.repositories.ProcessedDataRepository;
import pl.edu.agh.eaiib.io.sp.repositories.ReadingsRepository;
import pl.edu.agh.eaiib.io.sp.utils.EvaluatorHelper;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class ProcessorService {

    private final ProcessedDataRepository processedDataRepository;

    private final ReadingsRepository readingsRepository;

    @Autowired
    public ProcessorService(ProcessedDataRepository processedDataRepository, ReadingsRepository readingsRepository) {
        this.processedDataRepository = processedDataRepository;
        this.readingsRepository = readingsRepository;
    }

    public void clean() {
        processedDataRepository.deleteAll();
    }

    public void processData() {

    }

    public void processData(String deviceId, long start) {
        List<Reading> readingList = readingsRepository.findAll();
        Stream<Reading> readingStream = readingList.stream()
                .filter(x -> x.getDeviceId().equals(deviceId))
                .filter(x -> x.getTimestamp() >= start);

        Double min = readingStream.mapToDouble(x -> x.getAccelerometer().get(2)).min().orElse(0);
        Double max = readingStream.mapToDouble(x -> x.getAccelerometer().get(2)).max().orElse(20);

        List<ProcessedData> processedDataList =
                readingStream.map(x -> new ProcessedData(x.getLocation(), EvaluatorHelper.evaluate(x.getAccelerometer().get(2), min, max)))
                        .collect(Collectors.toList());

        processedDataRepository.saveAll(processedDataList);
    }
}

File: server/src/main/kotlin/pl/edu/agh/eaiib/io/sp/service/ProcessorService.kt
package pl.edu.agh.eaiib.io.sp.service

import org.springframework.stereotype.Service
import pl.edu.agh.eaiib.io.sp.analysis.DataAnalyzer
import pl.edu.agh.eaiib.io.sp.common.model.ProcessedData
import pl.edu.agh.eaiib.io.sp.repositories.ProcessedDataRepository
import pl.edu.agh.eaiib.io.sp.repositories.ReadingsRepository
import pl.edu.agh.eaiib.io.sp.utils.evaluate

@Service
class ProcessorService(private val processedDataRepository: ProcessedDataRepository,
                       private val readingsRepository: ReadingsRepository,
                       private val dataAnalyzer: DataAnalyzer) {


    fun clean() {
        processedDataRepository.deleteAll()
    }

    fun processData(deviceId: String, startTimestamp: Long, endTimestamp: Long) {
        val toProcess = readingsRepository.findAll()
                .asSequence()
                .filter { it.deviceId == deviceId }
                .filter { it.timestamp in startTimestamp..endTimestamp }

        val processed = dataAnalyzer.process(toProcess)
        processedDataRepository.saveAll(processed)
    }

    fun processData(deviceId: String, start: Long) {
        val readingList = readingsRepository.findAll()
        val readingSequence = readingList.asSequence()
                .filter { it.deviceId == deviceId }
                .filter { it.timestamp >= start }

        val min = readingSequence.map { x -> x.accelerometer[2] }.min() ?: 0.0
        val max = readingSequence.map { x -> x.accelerometer[2] }.max() ?: 20.0

        val processedDataList = readingSequence.map { x -> ProcessedData(x.location, evaluate(x.accelerometer[2], min, max)) }
                .toList()

        processedDataRepository.saveAll(processedDataList)
    }
}
